pkg/rungroup/rungroup.go

```diff
@@ -65,15 +73,39 @@ func (g *Group) Run() error {
 	level.Debug(g.logger).Log("msg", "received interrupt error from first actor -- shutting down other actors", "err", initialActorErr)
 
 	// Signal all actors to stop.
+	numActors := int64(len(g.actors))
+	interruptWait := semaphore.NewWeighted(numActors)
 	for _, a := range g.actors {
-		level.Debug(g.logger).Log("msg", "interrupting actor", "actor", a.name)
-		a.interrupt(initialActorErr.err)
+		interruptWait.Acquire(context.Background(), 1)
+		go func(a rungroupActor) {
+			defer interruptWait.Release(1)
+			level.Debug(g.logger).Log("msg", "interrupting actor", "actor", a.name)
+			a.interrupt(initialActorErr.err)
+			level.Debug(g.logger).Log("msg", "interrupt complete", "actor", a.name)
+		}(a)
 	}
 
-	// Wait for all actors to stop.
+	interruptCtx, interruptCancel := context.WithTimeout(context.Background(), interruptTimeout)
+	defer interruptCancel()
+
+	// Wait for interrupts to complete, but only until we hit our interruptCtx timeout
+	if err := interruptWait.Acquire(interruptCtx, numActors); err != nil {
+		level.Debug(g.logger).Log("msg", "timeout waiting for interrupts to complete, proceeding with shutdown", "err", err)
+	}
+
+	// Wait for all other actors to stop, but only until we hit our executeReturnTimeout
+	timeoutTimer := time.NewTimer(executeReturnTimeout)
+	defer timeoutTimer.Stop()
 	for i := 1; i < cap(errors); i++ {
-		e := <-errors
-		level.Debug(g.logger).Log("msg", "successfully interrupted actor", "actor", e.errorSourceName, "index", i)
+		select {
+		case <-timeoutTimer.C:
```

Comment:
I _think_ this case potentially leaks goroutines, which might leave a zombie process?